### ISSUE TITLE
Auto-updating document with tag search

### ISSUE THREAD

**Issue by cpopell** [kortina/vscode-markdown-notes#16]
Something that kind of builds on the backlink search request is there are a number of different plugins (Markdown All in One) that do auto-updating TOCs (which indicates to me the faculty for this is available).

What about being able to specify a tag (or general search term) at the top of the document, and then auto-update that document (as long as it's open) with the search results, so you can essentially have a Great Books style syntopicon for your search terms that doesn't require you to keep opening search pane?

**Comment by lukesmurray** [comment]
My feeling is that you really want a document to represent the search. For example if you have a tag for `#markdown-notes` you could create a document representing the tag containing all back references to `#markdown-notes`. 

The benefit of this is a normalized design. All your back references are in one place. You could use peek definition to view the search inline without cluttering the current document. You also only have to auto update one document so the implementation may be easier.[andy matuschak made a tool for this type of work](https://github.com/andymatuschak/note-link-janitor) which may be revant.

**Comment by cpopell** [comment]
Andy's tool is great, and so is your idea,  but I want to take it one step further - why limit it to tags? And why only one per document?

**Comment by lukesmurray** [comment]
Yes I think of tags and wiki links being on the same level. A tag is limited because it cannot contain spaces. But a wiki link can also have back references to it. The workflow in this case would be to have a set of wiki links and tags at the top of the page.

``[[writing inbox]] [[reading list]] [[journal]] #markdown-notes``

Each link represents a document. The document can contain a list of back references. You can go to the document using go to definition or view all references to the document using the view references. You can transclude the document using peek definition and transclude references using peek references.

**Comment by cpopell** [comment]
So let's actually take this a step further, because I don't think of tags and wiki links being on the same level, but I do think of tags as specially delineated words - 

A wiki link points at a specific file that does or should exist. This file may be broader than a specific concept.

A tag is a way to denote a specific word (no spaces, but dashes yes?) which denotes a position in an ontology (flat or otherwise) that is folder structure cross cutting

Any arbitrary selection of words could in theory be a tag, if you're not constrained to the # marker.

I guess that, for example, if I do a lot of writing around legacy (which I do), it would be useful to be able to easily "insert index on a common separated list of words" without having to have those all be tagged already - so I could put "immortality, legacy, inheritance" and it would look for both the tagged and untagged version of those words and build that index, without me having to go back and change things.

**Comment by lukesmurray** [comment]
I think your usage of tags differs from mine but thats totally fine. Either way I think that a references view would solve the issue. Here is an example of the references view from my own notes. The references are for my own concept of a wiki link but they could be used similarly for a tag across multiple levels of files.

![image](https://user-images.githubusercontent.com/34020210/81029253-3b34cd00-8e52-11ea-8fc4-1585901f85ec.png)

**Comment by cpopell** [comment]
Well the question is, is that a distinct document that can be viewed outside the context of the extension? Or is it just in that pane?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by lukesmurray** [comment]
just a pane viewed in the context of the extension.

**Comment by kortina** [comment]
OK, there is a little bit of discussion relevant to this thread on this PR: https://github.com/kortina/vscode-markdown-notes/pull/14

I think we are going to want every `[[wiki-link]]` to register as a reference to the `wiki-link.md` note, which will allow us to latch into the References vscode APIs.

But now I am also thinking we are going to want our own custom Tree View Explorer which shows 2 things:

1. A list of all the other notes/lines that reference the current note being edited
2. Possibly all of the other notes which share tags with the current note?

Let me clarify (2) a bit: suppose I am editing `test.md` with the following contents:

```# Test

#draft #test 

Some text ...
```

The Tree View might show something like:

```
references
    test2.md
       As I discussed in [[test]] ...
    test3.md 
       Another mention of [test]] ...
#draft #test 
    test4.md
        This has the tags #draft and #test 
#draft  
  test5.md
    This has just the tag #draft 
#test 
   test6.md
     This has just the tag #test 
```

I know for sure the references would be useful to see, but is showing the similarly tagged notes also useful?

**Comment by cpopell** [comment]
Well, you're looking at the treeview problem, but I think configurability is the solution here perhaps as I suggested above.

Multiplication of files may not be the solution - but if I could easily create a command to insert the dynamically updated result of a backlink search, I could do it myself. Or make the automation an option.

I'm heavily inspired by the syntopicon, so I'd love to be able to do something like

#writing #productivity
(list of all backlinks)

#writing #meditation
(list of all backlinks)

in the same file

**Comment by lukesmurray** [comment]
@kortina I like your solution. Minor point though. I think simply showing `#draft` and `#test` separately will be simpler. otherwise you'll run into combinatorics. 4 tags would create 15 sections in the references panel

**Comment by kortina** [comment]
Word, I have a few ideas for how to handle gracefully, but need to see them in the UI. Will probably try to refactor w remark ast before addressing, but this is high on the list.
